In [1]:
from datetime import datetime
import operator
import itertools


import csv
with open('texts.csv', 'r') as f:
    reader = csv.reader(f)
    texts = list(reader)

with open('calls.csv', 'r') as f:
    reader = csv.reader(f)
    calls = list(reader)
    

In [63]:
unique_tele_nums = set()
for call in calls:
    unique_tele_nums.add(call[0])
    unique_tele_nums.add(call[1])


In [64]:
for text in texts:
    unique_tele_nums.add(text[0])
    unique_tele_nums.add(text[1])

In [66]:
len(unique_tele_nums)

570

In [67]:
print(f'There are {len(unique_tele_nums)} different telephone numbers in the records.')

There are 570 different telephone numbers in the records.


In [68]:
phone_dict={}
for call in calls:
    phone_dict[call[0]] = phone_dict.get(call[0],0) + int(call[3])

In [70]:
 
phone_dict

479

In [71]:
for call in calls:
    phone_dict[call[1]] = phone_dict.get(call[1],1) + int(call[3])
    phone_dict.update()

In [73]:
phone_dict

551

In [74]:
max_time = max(phone_dict, key=phone_dict.get)


In [75]:
max_time

'(080)33251027'

In [76]:
all_values = phone_dict.values()

In [77]:
max_value = max(all_values)

In [78]:
max_value

90456

In [81]:
longest_time = max(phone_dict.items(), key=operator.itemgetter(1))

In [82]:
print("{0} spent the longest time, {1} seconds, on the phone during September 2016.".format(*longest_time))

(080)33251027 spent the longest time, 90456 seconds, on the phone during September 2016.


In [83]:
calls

[['78130 00821', '98453 94494', '01-09-2016 06:01:12', '186'],
 ['78298 91466', '(022)28952819', '01-09-2016 06:01:59', '2093'],
 ['97424 22395', '(022)47410783', '01-09-2016 06:03:51', '1975'],
 ['93427 40118', '(080)33118033', '01-09-2016 06:11:23', '1156'],
 ['90087 42537', '(080)35121497', '01-09-2016 06:17:26', '573'],
 ['97427 87999', '(04344)322628', '01-09-2016 06:19:28', '2751'],
 ['(080)45291968', '90365 06212', '01-09-2016 06:30:36', '9'],
 ['78132 18081', '77956 90632', '01-09-2016 06:39:03', '3043'],
 ['98453 46196', '94005 06213', '01-09-2016 06:40:20', '2457'],
 ['78290 99865', '89071 31755', '01-09-2016 06:46:56', '9'],
 ['(04344)228249', '(080)43901222', '01-09-2016 06:50:04', '2329'],
 ['(080)62164823', '74066 93594', '01-09-2016 06:52:07', '300'],
 ['(0821)6141380', '90366 69257', '01-09-2016 06:54:44', '2147'],
 ['98446 66723', '83019 53227', '01-09-2016 06:56:16', '129'],
 ['90088 09624', '93434 31551', '01-09-2016 06:57:44', '133'],
 ['93427 56500', '98447 12671',

In [84]:
#Find all of the area codes and mobile prefixes initiated by "(080)" area code
calls_by_080 = [call for call in calls if call[0].startswith('(080)')]

In [86]:
len(calls_by_080)

1080

In [87]:
# lets filter for fixed lines that begings with 0 called by "(080)" area code

fix_lines = [call for call in calls_by_080 if call[1].startswith("(0")]


In [88]:
fix_lines

[['(080)67362492', '(04344)316423', '01-09-2016 07:24:45', '2258'],
 ['(080)69245029', '(044)49481100', '01-09-2016 07:34:19', '9'],
 ['(080)67362492', '(04344)316423', '01-09-2016 09:39:50', '726'],
 ['(080)32255824', '(080)62164823', '01-09-2016 09:57:10', '12'],
 ['(080)29435303', '(080)43685310', '01-09-2016 10:25:39', '236'],
 ['(080)46772413', '(04344)211113', '01-09-2016 11:29:20', '2418'],
 ['(080)62164823', '(080)32255824', '01-09-2016 12:25:18', '3884'],
 ['(080)43215621', '(080)47459867', '01-09-2016 14:56:12', '695'],
 ['(080)43562014', '(040)34008657', '01-09-2016 21:03:36', '887'],
 ['(080)46566171', '(080)40395498', '01-09-2016 21:13:18', '938'],
 ['(080)20227149', '(080)39991213', '01-09-2016 22:19:58', '942'],
 ['(080)32390650', '(080)43562014', '01-09-2016 23:41:26', '1058'],
 ['(080)69245029', '(044)49481100', '02-09-2016 08:07:32', '103'],
 ['(080)43215621', '(080)47459867', '02-09-2016 08:32:17', '637'],
 ['(080)46772413', '(04344)211113', '02-09-2016 10:39:35', '2

In [91]:
# lets filter calling numbers starting with (080) from the calls into a list

all_fix_lines = [x[1].split(')') for x in fix_lines]

In [92]:
all_fix_lines

[['(04344', '316423'],
 ['(044', '49481100'],
 ['(04344', '316423'],
 ['(080', '62164823'],
 ['(080', '43685310'],
 ['(04344', '211113'],
 ['(080', '32255824'],
 ['(080', '47459867'],
 ['(040', '34008657'],
 ['(080', '40395498'],
 ['(080', '39991213'],
 ['(080', '43562014'],
 ['(044', '49481100'],
 ['(080', '47459867'],
 ['(04344', '211113'],
 ['(080', '39755879'],
 ['(040', '26738737'],
 ['(040', '34008657'],
 ['(080', '43562014'],
 ['(022', '22288051'],
 ['(080', '61123756'],
 ['(080', '44046839'],
 ['(04344', '211113'],
 ['(022', '22288051'],
 ['(022', '22288051'],
 ['(080', '32390650'],
 ['(080', '25863765'],
 ['(080', '47459867'],
 ['(022', '22288051'],
 ['(080', '44389098'],
 ['(080', '32255824'],
 ['(080', '61419142'],
 ['(080', '64047472'],
 ['(040', '34008657'],
 ['(080', '32255824'],
 ['(080', '22759842'],
 ['(080', '62164823'],
 ['(080', '22759842'],
 ['(080', '47459867'],
 ['(080', '26097534'],
 ['(080', '20227149'],
 ['(080', '31863188'],
 ['(04344', '211113'],
 ['(04546',

In [93]:
all_fix_lines_codes = [x[0] + x[0].join(')') for x in all_fix_lines]

In [130]:
mobile_list = [x[1] for x in calls_by_080 if x[1].startswith('7') or x[1].startswith('8') or x[1].startswith('9')]

In [131]:
mobile_list

['90365 06212',
 '74066 93594',
 '98440 65896',
 '98445 71741',
 '99003 88572',
 '97402 57057',
 '97380 86781',
 '81513 36123',
 '90088 09624',
 '98448 88411',
 '90351 35947',
 '95263 76972',
 '98445 71741',
 '84319 52539',
 '98448 88411',
 '98445 71741',
 '81513 36123',
 '99001 95783',
 '87146 31956',
 '92415 66985',
 '99001 95783',
 '93414 19669',
 '98442 73671',
 '97416 18084',
 '74064 33807',
 '90088 09624',
 '93426 76415',
 '83013 85044',
 '74066 93594',
 '97380 86781',
 '99612 41256',
 '98445 71741',
 '90194 08626',
 '78293 38561',
 '90199 67471',
 '93432 65750',
 '92415 66985',
 '93414 19669',
 '98447 62998',
 '93412 40385',
 '81517 73746',
 '97415 99182',
 '99001 95783',
 '78130 36804',
 '98440 65896',
 '98447 62998',
 '98445 27876',
 '90087 42537',
 '90199 67471',
 '98447 62998',
 '95266 42732',
 '81513 36123',
 '90199 67471',
 '84319 52539',
 '98440 65896',
 '90355 49499',
 '98458 94162',
 '84313 80377',
 '90351 35947',
 '81524 55552',
 '94001 62063',
 '98445 27876',
 '94005 

In [102]:
mobile_num_codes = [x[1][:4] for x in calls_by_080 if " " in x[1]]

In [103]:
mobile_num_codes

['9036',
 '7406',
 '9844',
 '9844',
 '9900',
 '9740',
 '9738',
 '8151',
 '9008',
 '9844',
 '9035',
 '9526',
 '9844',
 '8431',
 '9844',
 '9844',
 '8151',
 '9900',
 '8714',
 '9241',
 '9900',
 '9341',
 '9844',
 '9741',
 '7406',
 '9008',
 '9342',
 '8301',
 '7406',
 '9738',
 '9961',
 '9844',
 '9019',
 '7829',
 '9019',
 '9343',
 '9241',
 '9341',
 '9844',
 '9341',
 '8151',
 '9741',
 '9900',
 '7813',
 '9844',
 '9844',
 '9844',
 '9008',
 '9019',
 '9844',
 '9526',
 '8151',
 '9019',
 '8431',
 '9844',
 '9035',
 '9845',
 '8431',
 '9035',
 '8152',
 '9400',
 '9844',
 '9400',
 '9844',
 '8714',
 '9343',
 '8151',
 '9845',
 '8714',
 '9035',
 '7406',
 '9844',
 '9844',
 '7406',
 '7795',
 '9740',
 '9900',
 '9400',
 '9341',
 '9242',
 '9844',
 '9448',
 '9036',
 '9242',
 '9036',
 '9844',
 '9844',
 '9342',
 '9845',
 '9900',
 '9241',
 '9035',
 '7406',
 '9342',
 '8431',
 '9741',
 '9400',
 '7406',
 '9008',
 '9844',
 '9242',
 '7813',
 '9526',
 '9400',
 '8151',
 '9342',
 '9844',
 '9008',
 '9448',
 '8431',
 '9844',
 

In [115]:
all_codes = all_fix_lines_codes + mobile_num_codes

In [116]:
unique_codes = set()

In [117]:
for x in all_codes:
    unique_codes.add(x)

In [118]:
unique_codes

{'(022)',
 '(040)',
 '(04344)',
 '(044)',
 '(04546)',
 '(0471)',
 '(080)',
 '(0821)',
 '7406',
 '7795',
 '7813',
 '7829',
 '8151',
 '8152',
 '8301',
 '8431',
 '8714',
 '9008',
 '9019',
 '9035',
 '9036',
 '9241',
 '9242',
 '9341',
 '9342',
 '9343',
 '9400',
 '9448',
 '9449',
 '9526',
 '9656',
 '9738',
 '9740',
 '9741',
 '9742',
 '9844',
 '9845',
 '9900',
 '9961'}

The numbers called by people in Bangalore have codes: (022)
(040)
(04344)
(044)
(04546)
(0471)
(080)
(0821)
7406
7795
7813
7829
8151
8152
8301
8431
8714
9008
9019
9035
9036
9241
9242
9341
9342
9343
9400
9448
9449
9526
9656
9738
9740
9741
9742
9844
9845
9900
9961


In [119]:
print(f"The numbers called by people in Bangalore have codes:",'\n'.join(map(str, sorted(unique_codes)))) 


The numbers called by people in Bangalore have codes: (022)
(040)
(04344)
(044)
(04546)
(0471)
(080)
(0821)
7406
7795
7813
7829
8151
8152
8301
8431
8714
9008
9019
9035
9036
9241
9242
9341
9342
9343
9400
9448
9449
9526
9656
9738
9740
9741
9742
9844
9845
9900
9961


In [120]:
outgoing = set()
non_tele = set()

for call in calls:
    outgoing.add(call[0])
    non_tele.add(call[1])


In [121]:
for text in texts:
    outgoing.add(text[0])
    non_tele.add(text[1])


In [123]:
len(outgoing)

517

In [124]:
non_tele

{'(011)21017178',
 '(011)47486239',
 '(011)68486606',
 '(022)21884607',
 '(022)22288051',
 '(022)28765220',
 '(022)28952819',
 '(022)29303640',
 '(022)30044349',
 '(022)32517986',
 '(022)34715405',
 '(022)38214945',
 '(022)39006198',
 '(022)40840621',
 '(022)44927512',
 '(022)45444747',
 '(022)46574732',
 '(022)47410783',
 '(022)60273083',
 '(022)62120910',
 '(022)66911540',
 '(033)25441815',
 '(040)20512985',
 '(040)25325254',
 '(040)26738737',
 '(040)34008657',
 '(040)36649724',
 '(040)66729318',
 '(040)69695864',
 '(04344)211113',
 '(04344)212961',
 '(04344)228249',
 '(04344)316423',
 '(04344)322628',
 '(04344)615310',
 '(04344)617351',
 '(04344)649705',
 '(04344)654396',
 '(044)20550065',
 '(044)24037112',
 '(044)25144377',
 '(044)27523585',
 '(044)27641880',
 '(044)30360652',
 '(044)30727085',
 '(044)38044356',
 '(044)41581342',
 '(044)45416964',
 '(044)45838604',
 '(044)48154960',
 '(044)49481100',
 '(044)49868415',
 '(044)64005505',
 '(044)68407232',
 '(044)69775060',
 '(04546)2

In [125]:
tele = outgoing.difference(non_tele)

In [127]:
len(tele)

48

In [128]:
print("The numbers called by people in Bangalore have codes:",'\n'.join(map(str, sorted(tele)))) 


The numbers called by people in Bangalore have codes: (022)37572285
(022)65548497
(022)68535788
(022)69042431
(040)30429041
(044)22020822
(0471)2171438
(0471)6579079
(080)20383942
(080)25820765
(080)31606520
(080)40362016
(080)60463379
(080)60998034
(080)62963633
(080)64015211
(080)69887826
(0821)3257740
1400481538
1401747654
1402316533
1403072432
1403579926
1404073047
1404368883
1404787681
1407539117
1408371942
1408409918
1408672243
1409421631
1409668775
1409994233
74064 66270
74294 40351
78291 94593
87144 55014
87149 57915
90351 90193
90363 85738
92414 69419
94480 31371
94495 03761
97404 30456
97407 84573
97442 45192
99001 85033
99617 25274


In [2]:
times_stamp = [x[2] for x in calls] 

calling_number = [x[0] for x in calls]

receiving_number = [x[1] for x in calls]

duration_seconds = [x[3] for x in calls]


In [3]:
#converting timestamp by creating an empty list and appending coverted time to the the list 
times_stamp2 =[]
for x in times_stamp:
    x = datetime.strptime(x,'%d-%m-%Y %H:%M:%S')
    times_stamp2.append(x)

 
    
 

In [4]:
# converting duration into integers
duration_seconds2 =[]
for x in duration_seconds:
    x = int(x)
    duration_seconds2.append(x) 
        
    
    

In [5]:
call_records = []
for calling_num, receiving_num, times_stamp, duration_seconds in zip(calling_number, receiving_number, times_stamp2, duration_seconds2):
    record = {'calling_num':calling_num, 'receiving_num':receiving_num, 'times_stamp':times_stamp, 'duration_seconds':duration_seconds}
    call_records.append(record)

In [59]:
call_records[-1]

{'calling_num': '98447 62998',
 'receiving_num': '(080)46304537',
 'times_stamp': datetime.datetime(2016, 9, 30, 23, 57, 15),
 'duration_seconds': 2151}

In [62]:
print('First record of texts, {0} texts {1} at time {2}'.format(*texts[0]))
print('Last record of calls, {0} calls {1} at time {2}, lasting {3} seconds'.format(*calls[-1]))

First record of texts, 97424 22395 texts 90365 06212 at time 01-09-2016 06:03:22
Last record of calls, 98447 62998 calls (080)46304537 at time 30-09-2016 23:57:15, lasting 2151 seconds


In [7]:
# list for tex numbers
calling_num_text = [x[0] for x in texts] # O(n)

receiving_num_text = [x[1] for x in texts] # O(n)

times_text = [x[2] for x in texts] # O(n)

#converting text times to date time object
times_stamp_text =[]
for x in times_text:
    x = datetime.strptime(x,'%d-%m-%Y %H:%M:%S')
    times_stamp_text.append(x) # O(n)
    

In [8]:
text_records = []
for calling_num, receiving_num, times_stamp in zip(calling_num_text, receiving_num_text, times_stamp_text):
    record = {'calling_num':calling_num, 'receiving_num':receiving_num, 'times_stamp':times_stamp}
    text_records.append(record)

In [9]:
text_records

[{'calling_num': '97424 22395',
  'receiving_num': '90365 06212',
  'times_stamp': datetime.datetime(2016, 9, 1, 6, 3, 22)},
 {'calling_num': '94489 72078',
  'receiving_num': '92415 91418',
  'times_stamp': datetime.datetime(2016, 9, 1, 6, 5, 35)},
 {'calling_num': '81520 43406',
  'receiving_num': '92421 64236',
  'times_stamp': datetime.datetime(2016, 9, 1, 6, 9, 34)},
 {'calling_num': '97389 12538',
  'receiving_num': '90352 50054',
  'times_stamp': datetime.datetime(2016, 9, 1, 6, 9, 39)},
 {'calling_num': '81515 42171',
  'receiving_num': '98440 02823',
  'times_stamp': datetime.datetime(2016, 9, 1, 6, 13, 30)},
 {'calling_num': '78132 18081',
  'receiving_num': '77956 90632',
  'times_stamp': datetime.datetime(2016, 9, 1, 6, 19, 37)},
 {'calling_num': '90352 50054',
  'receiving_num': '97389 12538',
  'times_stamp': datetime.datetime(2016, 9, 1, 6, 30, 42)},
 {'calling_num': '90359 99240',
  'receiving_num': '94491 55790',
  'times_stamp': datetime.datetime(2016, 9, 1, 6, 32, 59

In [55]:
sorted_texts = sorted(text_records, key=operator.itemgetter('calling_num'),reverse=True)

In [58]:
sorted_calls[1]

{'calling_num': '(011)21017178',
 'receiving_num': '83015 22251',
 'times_stamp': datetime.datetime(2016, 9, 27, 11, 24, 59),
 'duration_seconds': 215}

In [10]:
sorted_calls = sorted(call_records, key=operator.itemgetter('calling_num'))

In [11]:
len(call_records)

5213

In [12]:
grouped_list = []
for key, items in itertools.groupby(sorted_calls, operator.itemgetter('calling_num')):
    
    grouped_list.append(list(items))

In [13]:
len(grouped_list)

479

In [14]:
new_filt = [x for x in sorted_calls if x['calling_num'].startswith('(080)')]

In [15]:
new_filt

[{'calling_num': '(080)20123809',
  'receiving_num': '93426 76415',
  'times_stamp': datetime.datetime(2016, 9, 2, 12, 28, 12),
  'duration_seconds': 3055},
 {'calling_num': '(080)20123809',
  'receiving_num': '93426 76415',
  'times_stamp': datetime.datetime(2016, 9, 4, 18, 24, 33),
  'duration_seconds': 95},
 {'calling_num': '(080)20123809',
  'receiving_num': '93426 76415',
  'times_stamp': datetime.datetime(2016, 9, 7, 6, 42, 46),
  'duration_seconds': 205},
 {'calling_num': '(080)20123809',
  'receiving_num': '93432 65750',
  'times_stamp': datetime.datetime(2016, 9, 12, 18, 49, 20),
  'duration_seconds': 1909},
 {'calling_num': '(080)20123809',
  'receiving_num': '93426 76415',
  'times_stamp': datetime.datetime(2016, 9, 13, 22, 53, 32),
  'duration_seconds': 630},
 {'calling_num': '(080)20123809',
  'receiving_num': '93426 76415',
  'times_stamp': datetime.datetime(2016, 9, 16, 18, 37, 27),
  'duration_seconds': 1092},
 {'calling_num': '(080)20123809',
  'receiving_num': '93426 

In [16]:
fix_num = [x for x in new_filt if x['receiving_num'].startswith("(0")]

In [17]:
fix_num

[{'calling_num': '(080)20227149',
  'receiving_num': '(080)39991213',
  'times_stamp': datetime.datetime(2016, 9, 1, 22, 19, 58),
  'duration_seconds': 942},
 {'calling_num': '(080)20227149',
  'receiving_num': '(080)64047472',
  'times_stamp': datetime.datetime(2016, 9, 3, 15, 10, 30),
  'duration_seconds': 202},
 {'calling_num': '(080)20227149',
  'receiving_num': '(080)39991213',
  'times_stamp': datetime.datetime(2016, 9, 4, 21, 4, 6),
  'duration_seconds': 2424},
 {'calling_num': '(080)20227149',
  'receiving_num': '(080)64047472',
  'times_stamp': datetime.datetime(2016, 9, 13, 11, 11, 57),
  'duration_seconds': 2925},
 {'calling_num': '(080)20227149',
  'receiving_num': '(080)64047472',
  'times_stamp': datetime.datetime(2016, 9, 15, 13, 31, 5),
  'duration_seconds': 2122},
 {'calling_num': '(080)20227149',
  'receiving_num': '(080)39991213',
  'times_stamp': datetime.datetime(2016, 9, 16, 13, 18, 58),
  'duration_seconds': 9},
 {'calling_num': '(080)20227149',
  'receiving_num'

In [18]:
all_fix_num = 'receiving_num'
all_fix = [key['receiving_num'] for key in fix_num]

In [19]:
all_fix

['(080)39991213',
 '(080)64047472',
 '(080)39991213',
 '(080)64047472',
 '(080)64047472',
 '(080)39991213',
 '(080)39991213',
 '(080)39991213',
 '(080)39991213',
 '(080)39991213',
 '(080)39991213',
 '(080)39991213',
 '(080)64047472',
 '(040)26738737',
 '(040)26738737',
 '(040)26738737',
 '(040)26738737',
 '(040)26738737',
 '(040)26738737',
 '(040)26738737',
 '(040)26738737',
 '(040)26738737',
 '(080)43562014',
 '(04546)267875',
 '(080)43562014',
 '(080)43562014',
 '(080)43562014',
 '(080)43562014',
 '(080)43562014',
 '(04546)267875',
 '(080)43562014',
 '(04546)267875',
 '(080)43562014',
 '(080)43562014',
 '(080)43562014',
 '(080)43562014',
 '(080)43562014',
 '(080)43562014',
 '(0821)6141380',
 '(0821)6141380',
 '(0821)6141380',
 '(0821)6141380',
 '(0821)6141380',
 '(0821)6141380',
 '(0821)6141380',
 '(0821)6141380',
 '(0821)6141380',
 '(0821)6141380',
 '(080)23413513',
 '(080)23413513',
 '(080)23413513',
 '(022)38214945',
 '(022)28952819',
 '(022)38214945',
 '(022)38214945',
 '(022)669

In [20]:
all_fix_area_Code = [x.split(')') for x in all_fix]

In [21]:
all_fix_area_Code

[['(080', '39991213'],
 ['(080', '64047472'],
 ['(080', '39991213'],
 ['(080', '64047472'],
 ['(080', '64047472'],
 ['(080', '39991213'],
 ['(080', '39991213'],
 ['(080', '39991213'],
 ['(080', '39991213'],
 ['(080', '39991213'],
 ['(080', '39991213'],
 ['(080', '39991213'],
 ['(080', '64047472'],
 ['(040', '26738737'],
 ['(040', '26738737'],
 ['(040', '26738737'],
 ['(040', '26738737'],
 ['(040', '26738737'],
 ['(040', '26738737'],
 ['(040', '26738737'],
 ['(040', '26738737'],
 ['(040', '26738737'],
 ['(080', '43562014'],
 ['(04546', '267875'],
 ['(080', '43562014'],
 ['(080', '43562014'],
 ['(080', '43562014'],
 ['(080', '43562014'],
 ['(080', '43562014'],
 ['(04546', '267875'],
 ['(080', '43562014'],
 ['(04546', '267875'],
 ['(080', '43562014'],
 ['(080', '43562014'],
 ['(080', '43562014'],
 ['(080', '43562014'],
 ['(080', '43562014'],
 ['(080', '43562014'],
 ['(0821', '6141380'],
 ['(0821', '6141380'],
 ['(0821', '6141380'],
 ['(0821', '6141380'],
 ['(0821', '6141380'],
 ['(0821', 

In [22]:
all_fix_codes = [x[0] + x[0].join(')') for x in all_fix_area_Code]

In [23]:
all_fix_codes

['(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(040)',
 '(040)',
 '(040)',
 '(040)',
 '(040)',
 '(040)',
 '(040)',
 '(040)',
 '(040)',
 '(080)',
 '(04546)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(04546)',
 '(080)',
 '(04546)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(0821)',
 '(0821)',
 '(0821)',
 '(0821)',
 '(0821)',
 '(0821)',
 '(0821)',
 '(0821)',
 '(0821)',
 '(0821)',
 '(080)',
 '(080)',
 '(080)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(022)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(080)',
 '(0

In [24]:
mobile = [x for x in new_filt if " " in x['receiving_num']]

In [25]:
mobile

[{'calling_num': '(080)20123809',
  'receiving_num': '93426 76415',
  'times_stamp': datetime.datetime(2016, 9, 2, 12, 28, 12),
  'duration_seconds': 3055},
 {'calling_num': '(080)20123809',
  'receiving_num': '93426 76415',
  'times_stamp': datetime.datetime(2016, 9, 4, 18, 24, 33),
  'duration_seconds': 95},
 {'calling_num': '(080)20123809',
  'receiving_num': '93426 76415',
  'times_stamp': datetime.datetime(2016, 9, 7, 6, 42, 46),
  'duration_seconds': 205},
 {'calling_num': '(080)20123809',
  'receiving_num': '93432 65750',
  'times_stamp': datetime.datetime(2016, 9, 12, 18, 49, 20),
  'duration_seconds': 1909},
 {'calling_num': '(080)20123809',
  'receiving_num': '93426 76415',
  'times_stamp': datetime.datetime(2016, 9, 13, 22, 53, 32),
  'duration_seconds': 630},
 {'calling_num': '(080)20123809',
  'receiving_num': '93426 76415',
  'times_stamp': datetime.datetime(2016, 9, 16, 18, 37, 27),
  'duration_seconds': 1092},
 {'calling_num': '(080)20123809',
  'receiving_num': '93426 

In [26]:
all_mobile_num = 'receiving_num'
all_mobile_phones = [key['receiving_num'] for key in mobile]
 

In [27]:
all_mobile_phones

['93426 76415',
 '93426 76415',
 '93426 76415',
 '93432 65750',
 '93426 76415',
 '93426 76415',
 '93426 76415',
 '93426 76415',
 '93426 76415',
 '93432 65750',
 '93432 65750',
 '81517 73746',
 '81517 73746',
 '81517 73746',
 '81517 73746',
 '81517 73746',
 '81517 73746',
 '81517 73746',
 '78293 38561',
 '78293 38561',
 '78293 38561',
 '78293 38561',
 '78293 38561',
 '78293 38561',
 '78293 38561',
 '94488 97308',
 '94488 97308',
 '94488 97308',
 '94488 97308',
 '94488 97308',
 '94488 97308',
 '94488 97308',
 '94488 97308',
 '94488 97308',
 '94488 97308',
 '94488 97308',
 '90088 09624',
 '90088 09624',
 '90088 09624',
 '90088 09624',
 '90088 09624',
 '90088 09624',
 '90088 09624',
 '90088 09624',
 '90088 09624',
 '90088 09624',
 '98445 71741',
 '98445 71741',
 '98445 71741',
 '98445 71741',
 '98445 71741',
 '98445 71741',
 '98445 71741',
 '98445 71741',
 '98445 71741',
 '98445 71741',
 '94491 55790',
 '92423 51078',
 '92423 51078',
 '92423 51078',
 '92423 51078',
 '92423 51078',
 '98440 

In [28]:
mobile_code_list =[x[:4] for x in all_mobile_phones if x.startswith('7') or x.startswith('8') or x.startswith('9')]

In [29]:
mobile_code_list

['9342',
 '9342',
 '9342',
 '9343',
 '9342',
 '9342',
 '9342',
 '9342',
 '9342',
 '9343',
 '9343',
 '8151',
 '8151',
 '8151',
 '8151',
 '8151',
 '8151',
 '8151',
 '7829',
 '7829',
 '7829',
 '7829',
 '7829',
 '7829',
 '7829',
 '9448',
 '9448',
 '9448',
 '9448',
 '9448',
 '9448',
 '9448',
 '9448',
 '9448',
 '9448',
 '9448',
 '9008',
 '9008',
 '9008',
 '9008',
 '9008',
 '9008',
 '9008',
 '9008',
 '9008',
 '9008',
 '9844',
 '9844',
 '9844',
 '9844',
 '9844',
 '9844',
 '9844',
 '9844',
 '9844',
 '9844',
 '9449',
 '9242',
 '9242',
 '9242',
 '9242',
 '9242',
 '9844',
 '9844',
 '9844',
 '9844',
 '9844',
 '9844',
 '9036',
 '9036',
 '9036',
 '9036',
 '9036',
 '9036',
 '9036',
 '8151',
 '8151',
 '8151',
 '8151',
 '8151',
 '8151',
 '9738',
 '9738',
 '9738',
 '9738',
 '9738',
 '9738',
 '9738',
 '9738',
 '9738',
 '9738',
 '9738',
 '9738',
 '9738',
 '9738',
 '9900',
 '9900',
 '9900',
 '9900',
 '9900',
 '9900',
 '9900',
 '9343',
 '9526',
 '9343',
 '9526',
 '9343',
 '9844',
 '9844',
 '9844',
 '9844',
 

In [30]:
all_area_codes = all_fix_codes + mobile_code_list

In [31]:
me = sorted(all_area_codes)

In [32]:
mee =set(me)

In [109]:
len(mee)

39

In [34]:
def unique(x):
 
    # initialize a null list
    unique_list = []
     
    # traverse for all elements
    for num in x:
        # check if exists in unique_list or not
        if num not in unique_list:
            unique_list.append(num)
    # print list
    print("The numbers called by people in Bangalore have codes:",'\n'.join(map(str, sorted(unique_list)))) 
    


In [111]:
print(unique(len(all_area_codes)))

TypeError: 'int' object is not iterable

In [36]:
# PART B

#Lets find the percentage of calls made from a number starting with "(080)" to a number also starting with "(080)"
# The  list new_filt contains all records of calls initiated by (080)
#lets filter further to extract the receiving number starting with (080) and stor it in fix2 list
fix2 = [x for x in new_filt if x['receiving_num'].startswith("(080")]

#the percentage of calls is expressed as the len of fix2 / the len of the entire calls dataframe

print(round(len(fix2) / len(sorted_calls) * 100, 2),'percent of calls from fixed lines in Bangalore are calls to other fixed lines in Bangalore.')


5.14 percent of calls from fixed lines in Bangalore are calls to other fixed lines in Bangalore.


nuber4

In [37]:
marketers = [x for x in sorted_calls if x['calling_num'].startswith('140')]

In [38]:
marketers

[{'calling_num': '1400481538',
  'receiving_num': '(022)29303640',
  'times_stamp': datetime.datetime(2016, 9, 2, 13, 32, 55),
  'duration_seconds': 24},
 {'calling_num': '1400481538',
  'receiving_num': '(08214175)358',
  'times_stamp': datetime.datetime(2016, 9, 4, 12, 49, 16),
  'duration_seconds': 98},
 {'calling_num': '1400481538',
  'receiving_num': '(080)41095396',
  'times_stamp': datetime.datetime(2016, 9, 4, 19, 43, 43),
  'duration_seconds': 73},
 {'calling_num': '1400481538',
  'receiving_num': '(080)32828889',
  'times_stamp': datetime.datetime(2016, 9, 10, 17, 20, 21),
  'duration_seconds': 881},
 {'calling_num': '1400481538',
  'receiving_num': '(080)69104549',
  'times_stamp': datetime.datetime(2016, 9, 12, 18, 9, 14),
  'duration_seconds': 10},
 {'calling_num': '1400481538',
  'receiving_num': '97420 86635',
  'times_stamp': datetime.datetime(2016, 9, 19, 12, 0, 59),
  'duration_seconds': 73},
 {'calling_num': '1400481538',
  'receiving_num': '(044)25144377',
  'times_

In [41]:
receiving_calls_140 = [x for x in sorted_calls if x['receiving_num'].startswith('140')] # O(n)

#receiving_calls_140 # No records found in the receiving key of dictionaries in the calls list

#checking if any number starting with 140 is sending text

sending_text_140 =[x for x in sorted_texts if x['calling_num'].startswith('140')] # O(n)

#sending_text_140 #No records found for this category as well

#checking if any number starting with 140 is Receiving text

receiving_text_140 = [x for x in sorted_texts if x['receiving_num'].startswith('140')] # O(n)

#receiving_text_140 # No record found for this category 

# Getting the list with all numbers starting with 140 into a list named all_marketers 


In [42]:
receiving_calls_140

[]

In [43]:
sending_text_140

[]

In [44]:
receiving_text_140

[]

In [46]:
def unique2(x):
 
    # initialize a null list
    unique_list = []
     
    # traverse for all elements
    for num in x:
        # check if exists in unique_list or not
        if num not in unique_list:
            unique_list.append(num)
    # print list
    print("These numbers could be telemarketers: ",'\n'.join(map(str, sorted(unique_list)))) 
    


TypeError: '<' not supported between instances of 'dict' and 'dict'

In [48]:
 
all_marketers = [key['calling_num'] for key in marketers]

In [50]:
unique2(all_marketers)

These numbers could be telemarketers:  1400481538
1401747654
1402316533
1403072432
1403579926
1404073047
1404368883
1404787681
1407539117
1408371942
1408409918
1408672243
1409421631
1409668775
1409994233


In [49]:
all_marketers

['1400481538',
 '1400481538',
 '1400481538',
 '1400481538',
 '1400481538',
 '1400481538',
 '1400481538',
 '1400481538',
 '1400481538',
 '1401747654',
 '1401747654',
 '1401747654',
 '1401747654',
 '1402316533',
 '1402316533',
 '1402316533',
 '1402316533',
 '1402316533',
 '1402316533',
 '1402316533',
 '1402316533',
 '1402316533',
 '1402316533',
 '1402316533',
 '1403072432',
 '1403072432',
 '1403072432',
 '1403072432',
 '1403072432',
 '1403072432',
 '1403579926',
 '1403579926',
 '1403579926',
 '1404073047',
 '1404073047',
 '1404073047',
 '1404073047',
 '1404073047',
 '1404073047',
 '1404073047',
 '1404368883',
 '1404368883',
 '1404787681',
 '1404787681',
 '1404787681',
 '1407539117',
 '1407539117',
 '1407539117',
 '1407539117',
 '1407539117',
 '1407539117',
 '1407539117',
 '1407539117',
 '1407539117',
 '1408371942',
 '1408371942',
 '1408371942',
 '1408371942',
 '1408371942',
 '1408371942',
 '1408371942',
 '1408371942',
 '1408371942',
 '1408371942',
 '1408371942',
 '1408371942',
 '14084099